File: src/Scanner.Core/Checks/InvalidAriaRoleCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

public sealed class InvalidAriaRoleCheck : ICheck
{
    private static readonly HashSet<string> AllowedRoles = new(StringComparer.OrdinalIgnoreCase)
    {
        "banner",
        "button",
        "checkbox",
        "dialog",
        "grid",
        "link",
        "list",
        "listitem",
        "main",
        "navigation",
        "region",
        "search",
        "textbox"
    };

    private static readonly Regex RoleRegex = new("role=\"(?<role>[^\"]+)\"", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    public string Id => "invalid-aria-role";
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        foreach (Match match in RoleRegex.Matches(context.Content))
        {
            var role = match.Groups["role"].Value;
            if (AllowedRoles.Contains(role))
            {
                continue;
            }

            var line = TextUtilities.GetLineNumber(context.Content, match.Index);
            yield return new Issue(rule.Id, Id, context.FilePath, line, $"Invalid ARIA role '{role}'.", match.Value);
        }
    }
}

File: node_engine/src/checks/InvalidAriaRoleCheck.js
const { getLineNumber } = require("./TextUtilities");

const allowedRoles = new Set([
  "alert",
  "alertdialog",
  "application",
  "article",
  "banner",
  "button",
  "cell",
  "checkbox",
  "columnheader",
  "combobox",
  "complementary",
  "contentinfo",
  "definition",
  "dialog",
  "directory",
  "document",
  "feed",
  "figure",
  "form",
  "grid",
  "gridcell",
  "group",
  "heading",
  "img",
  "link",
  "list",
  "listbox",
  "listitem",
  "log",
  "main",
  "marquee",
  "math",
  "menu",
  "menubar",
  "menuitem",
  "menuitemcheckbox",
  "menuitemradio",
  "navigation",
  "none",
  "note",
  "option",
  "presentation",
  "progressbar",
  "radio",
  "radiogroup",
  "region",
  "row",
  "rowgroup",
  "rowheader",
  "scrollbar",
  "search",
  "searchbox",
  "separator",
  "slider",
  "spinbutton",
  "status",
  "switch",
  "tab",
  "table",
  "tablist",
  "tabpanel",
  "term",
  "textbox",
  "timer",
  "toolbar",
  "tooltip",
  "tree",
  "treegrid",
  "treeitem"
]);

const roleRegex = /role="(?<role>[^"]+)"/gi;

const InvalidAriaRoleCheck = {
  id: "invalid-aria-role",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    const issues = [];
    for (const match of context.content.matchAll(roleRegex)) {
      const roleValue = match.groups.role;
      const tokens = roleValue
        .split(/\s+/)
        .map((token) => token.trim())
        .filter(Boolean);

      for (const token of tokens) {
        if (allowedRoles.has(token.toLowerCase())) {
          continue;
        }

        const line = getLineNumber(context.content, match.index);
        issues.push({
          ruleId: rule.id,
          checkId: InvalidAriaRoleCheck.id,
          filePath: context.filePath,
          line,
          message: `Invalid ARIA role '${token}'.`,
          evidence: match[0]
        });
      }
    }

    return issues;
  }
};

module.exports = { InvalidAriaRoleCheck, allowedRoles };
